src/bitmaps.rs

```diff
@@ -0,0 +1,199 @@
+use std::fs::File;
+use std::io::BufWriter;
+use std::path::Path;
+use std::{fs, io};
+
+use endio_bit::BitReader;
+
+use allsorts::binary::read::ReadScope;
+use allsorts::bitmap::{self, BitDepth, BitmapSize, CBDTTable, CBLCTable, GlyphBitmapData};
+use allsorts::fontfile::FontFile;
+use allsorts::tables::FontTableProvider;
+use allsorts::tag::{self};
+
+use crate::cli::BitmapOpts;
+use crate::BoxError;
+
+pub fn main(opts: BitmapOpts) -> Result<i32, BoxError> {
+    let buffer = std::fs::read(&opts.font)?;
+    let scope = ReadScope::new(&buffer);
+    let font_file = scope.read::<FontFile>()?;
+    let table_provider = font_file.table_provider(opts.index)?;
+
+    let table = table_provider
+        .table_data(tag::CBLC)?
+        .or(table_provider.table_data(tag::EBLC)?)
+        .ok_or("font does not have CBLC or EBLC tables")?;
+    let scope = ReadScope::new(table.borrow());
+    let cblc = scope.read::<CBLCTable<'_>>().unwrap();
+    let table = table_provider
+        .table_data(tag::CBDT)?
+        .or(table_provider.table_data(tag::EBDT)?)
+        .ok_or("font does not have CBDT or EBDT tables")?;
+    let scope = ReadScope::new(table.borrow());
+    let cbdt = scope.read::<CBDTTable<'_>>().unwrap();
+
+    let output_path = Path::new(&opts.output);
+    if !output_path.exists() {
+        fs::create_dir(output_path)?;
+    }
+
+    for strike in cblc.bitmap_sizes.iter_res() {
+        let strike = strike?;
+        let strike_path = output_path.join(&format!(
+            "{}x{}@{}",
+            strike.ppem_x, strike.ppem_y, strike.bit_depth as u8
+        ));
+        if !strike_path.exists() {
+            fs::create_dir(&strike_path)?;
+        }
+        dump_bitmaps(strike, &strike_path, &cblc, &cbdt)?;
+    }
+
+    Ok(0)
+}
+
+fn dump_bitmaps<'a>(
+    strike: BitmapSize,
+    path: &Path,
+    cblc: &'a CBLCTable<'a>,
+    cbdt: &'a CBDTTable<'a>,
+) -> Result<(), BoxError> {
+    for glyph_id in strike.start_glyph_index..=strike.end_glyph_index {
+        if let Some(bitmap) =
+            bitmap::lookup(glyph_id, strike.ppem_x, BitDepth::ThirtyTwo, cblc, cbdt)?
+        {
+            let glyph_path = path.join(&format!("{}.png", glyph_id));
+            match bitmap {
+                (_, GlyphBitmapData::Format17 { data, .. })
+                | (_, GlyphBitmapData::Format18 { data, .. })
+                | (_, GlyphBitmapData::Format19 { data, .. }) => {
+                    // Already PNG, just write it out
+                    fs::write(&glyph_path, data)?;
+                }
+                (_, GlyphBitmapData::Format8 { components, .. })
+                | (_, GlyphBitmapData::Format9 { components, .. }) => {
+                    let ids = components
+                        .iter()
+                        .map(|component| component.glyph_id.to_string())
+                        .collect::<Vec<_>>();
+                    println!("glyph {} is comprised of {}", glyph_id, ids.join(", "))
+                }
```

Comment:
Sorry, I don't know what these formats are. Are we ignoring them rather than `write`ing?

Reply:
Those ones are "compenent: formats. They say that this glyph is comprised of these other glyphs positioned at these coordinates. There's no new bitmap contained in them to dump. https://docs.microsoft.com/en-us/typography/opentype/spec/ebdt#ebdtcomponent-record

Reply:
Cool. Makes sense